### PR TITLE
Implement read/write, with deadline handling.

### PR DESCRIPTION
**Pull request by danderson**
With this change, pppoe.Conn implements net.Conn.

### CODE REVIEW

File: internal/pppoe/pppoe.go [hunk 1]
```diff
@@ -117,3 +124,29 @@ func (c *Conn) Close() error {
...
+	n, _, err := readSessionPacket(c.sessionFd, b, c.readDeadline)
+	return n, err
+}
+
+func (c *Conn) Write(b []byte) (int, error) {
+	// TODO: deadline
+	return sendSessionPacket(c.sessionFd, b, c.writeDeadline)
+}
+
+func (c *Conn) SetDeadline(deadline time.Time) error {
```
Comment (bradfitz) [inline internal/pppoe/pppoe.go:142]: Technically these three methods don't implement the net.Conn interface rules faithfully, which require that:

```go
// SetReadDeadline sets the deadline for future Read calls
// and any currently-blocked Read call.
// A zero value for t means Read will not time out.
```

Note the "and any currently-blocked Read call" part.

net/http depends on that, for instance.

It's checked by https://godoc.org/golang.org/x/net/nettest#TestConn too.

Comment (bradfitz) [reply internal/pppoe/pppoe.go:142]: So maybe you just want to document these as incomplete for now in the godoc and in the body a `TODO: if anybody ever cares`.

File: internal/pppoe/session.go [hunk 2]
```diff
@@ -26,17 +28,52 @@ func connectSessionFd(fd int, ifName string, remote net.HardwareAddr, sessionID
...
+	}
 	return unix.Recvfrom(fd, buf, 0)
 }
+
+func setTimeout(fd int, opt int, deadline time.Time) error {
+	var tv unix.Timeval
+	if !deadline.IsZero() {
+		d := time.Until(deadline).Truncate(time.Microsecond)
+		if d < 0 {
+			return errors.New("timeout") // TODO: net.Error
```
Comment (bradfitz) [inline internal/pppoe/session.go]: easy enough I'd do now. For instance, x/net/http2 contains:

```go
type httpError struct {
        msg     string
        timeout bool
}

func (e *httpError) Error() string   { return e.msg }
func (e *httpError) Timeout() bool   { return e.timeout }
func (e *httpError) Temporary() bool { return true }
```

Make something like that, or copy that.

**Review by bradfitz — APPROVED** [review]

File: internal/pppoe/pppoe.go (hunk 1 above)
Comment (danderson) [reply internal/pppoe/pppoe.go:142]: Done.

My expectation is that this won't get fixed until/unless the runtime's netpoller gets exposed. At which point I can just hook all of this into the main netpoller and have it correctly implement full timeout semantics for me with its more complex logic.

File: internal/pppoe/session.go (hunk 2 above)
Comment (danderson) [reply internal/pppoe/session.go]: Done.

**Pull request merged**